cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/AppResourceWatcherThread.java

```diff
@@ -95,13 +105,62 @@ Cancellable addListener(ResourceChangeListener<T> listener) {
   protected Watchable<T> createWatchable(Type resourceType, String namespace,
                                          @Nullable String labelSelector) throws IOException, ApiException {
     ApiClient apiClient = getApiClient();
-    CustomObjectsApi api = new CustomObjectsApi(apiClient);
-    Call call = api.listNamespacedCustomObjectCall(group, version, namespace, plural, null, null, null,
-                                                   labelSelector, null, null, null, true, null);
+    Call call;
+    if (resourceType.equals(V1Job.class)) {
+      call = getCallForJobList(namespace, labelSelector);
+    } else if (resourceType.equals(V1StatefulSet.class)) {
+      call = getCallForStatefulSetList(namespace, labelSelector);
+    } else if (resourceType.equals(V1Deployment.class)) {
+      call = getCallForDeploymentList(namespace, labelSelector);
+    } else {
+      CustomObjectsApi api = new CustomObjectsApi(apiClient);
+      call = api.listNamespacedCustomObjectCall(group, version, namespace, plural, null, null, null,
+                                                labelSelector, null, null, null, true, null);
+    }
     return Watch.createWatch(apiClient, call, TypeToken.getParameterized(Watch.Response.class,
                                                                          resourceType).getType());
   }
 
+  private Call getCallForJobList(String namespace, @Nullable String labelSelector) throws IOException, ApiException {
+    ApiClient apiClient = getApiClient();
+    BatchV1Api api = new BatchV1Api(apiClient);
+    if (resourceVersion == null) {
+      V1JobList jobList = api.listNamespacedJob(namespace, null, null, null, null, labelSelector,
+                                                       null, null, null, null, null);
+      resourceVersion = jobList.getMetadata().getResourceVersion();
+    }
+    return api.listNamespacedJobCall(namespace, null, null, null, null, labelSelector,
+                                            null, resourceVersion, null, null, true, null);
```

Comment:
ResourceVersionMatch is not set as it's forbidden for watch - https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/internalversion/validation/validation.go#L30.
Documentation indicates that watch starts at the exact resource version if it's set - https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-watch